Create Savi Java-TzWhere from prototype.

dsbos opened the pull request:
@jamescnowell @JimHaughwout

dsbos committed:
Narrowed .gitnore's exclusions.

Removed *.class, *.jar, etc., since they should be covered by ignoring of Maven
build directory "target".

Narrowed exclusion "target" to "/target" since this repository doesn't have
nested Maven projects.

dsbos committed:
Added ignoring of Emacs backup files ("*~").

dsbos committed:
Renamed org.sensoranalytics to com.savi.timezones.

dsbos committed:
Reorganized, reworked.  (Not cleaned yet.)

dsbos committed:
Added some test cases, cleaned.

dsbos committed:
Various cleanup.  Renamed package.  Cleaned up tests.  Etc.

Moved from com.savi.timezones to com.savi.geo.timezones.
Added some Eclipse file patterns to .gitignore.

dsbos committed:
Deleted directory "./Input data/" (see ./src/main/resources/).

dsbos committed:
Edited resources read-me file.

dsbos committed:
Updated REVIEW question comments.

dsbos committed:
Fixed typo.

jamescnowell commented on src/main/java/com/savi/geo/timezones/IndexImpl.java:
```diff
@@ -0,0 +1,191 @@
+    private final STRtree pointToTzApproxIndex;
+
+
+    //???? TODO: Probably either change from Feature to Geometry or change to get zone ID out of feature.
```
Looking at the code below, looks like you're right. Just the Geom is fine.

dsbos replied:
Oops. Missed.

Done

jamescnowell commented on src/main/resources/com/savi/geo/timezones/README.txt:
```diff
@@ -0,0 +1,2 @@
+The shapefile files tz_world.shp, tz_world.shx, and tz_world.dbf are taken from
```
Just as a question: did you grab new versions of the files? The diff shows just a rename with no changes, and I'm not sure if we started with the most up to date stuff.

dsbos replied:
Got new versions today.

jamescnowell commented on src/test/java/com/savi/geo/timezones/TzWhereTest.java:
```diff
@@ -0,0 +1,126 @@
+        // A point that is vertex on boundary polygon of Europe/Vatican:
+        // (Attempted exact values--but unconfirmed.)
+        final double latitude  = 41.9063606262207;
+        final double longitude = 12.45091629028328;
```
Wow, that's super precise. As to your statement in `getLatLongTimeZoneId`, I think we just need to be clear about what we expect to be able to provide. I think if we somehow hit a point that is legitimately inside two geometries, returning the last one is a sensible as anything else. It's going to be very hard to test all of those cases, and it will likely be easier to let reported issues dictate how we handle it in a more graceful way.

dsbos replied:
> Wow, that's super precise.
> It's just the same digits that got printed out when dumping the numbers. I don't recall if Java guarantees enough digits on output to get the same bits upon input.  It's not confirmed that the bits actually were the same, because it's not confirmed that two zones matched, but that might have been because Rome's boundary with Vatican City is a negative polygon, not a regular polygon.

Changed assertion to allow Rome too.

jamescnowell commented on src/main/java/com/savi/geo/timezones/TzWhere.java:
```diff
@@ -0,0 +1,61 @@
+//?? TODO:  REVIEW:  How much flexibility should we support right now?
```
I'd say let someone ask for it before we spend time working on it.

dsbos replied:
Okay.  Removed comment.

jamescnowell commented on src/main/java/com/savi/geo/timezones/TzWhere.java:
```diff
@@ -0,0 +1,61 @@
+//   with one static replacement (probably by defining system property used to
+//   override default classpath resource pathname to shapefile files)?
+
+//?? TODO:  REVIEW:  Should proper-noun--based class name "TzWhere" be changed
```
The more I think about it, the more I'm ok using TzWhere. If people have used the Python or Javascript versions, they know what they're getting here.

dsbos replied:
Okay.  Leaving it TzWhere.  Purged question comments.

jamescnowell commented on Input data/tz_world.prj:
```diff
@@ -1 +0,0 @@
-GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]]
```
We might want to keep the .prj file. It's not necessary for what we're doing, but it does indicate to anyone inspecting the package what ellpsoid is used to project the shapes.

dsbos replied:
Added .prj file.

jamescnowell commented on pom.xml:
```diff
@@ -4,17 +4,25 @@
 
-    <groupId>com.savi</groupId>
-    <artifactId>shapeconvert</artifactId>
+    <groupId>com.savi.nameThisWhichGroupId</groupId>
```
com.savi.geo

dsbos replied:
Done.

jamescnowell commented:
Aside from the line comments, as you noted, I think it's just a matter of deciding how to handle exceptions and weird edge cases. I say handle them however we want it to work for ourselves (which should be pretty simple, right?) and let other people file issues if they have a different expectation.

dsbos committed:
Re review: Set Maven artifact group ID to "com.savi.geo".

dsbos committed:
Re review:  Purged questions-for-review comments (most).

dsbos committed:
Fixed exactly-on-boundary case and test.

Fixed exactly-on-boundary test to actually be exact.
Fixed containment check to consider point on boundary to be in region.

dsbos committed:
Simplified FeatureAndId to GeometryAndId.

dsbos committed:
Got fresh copies of tz_world shapefile; added .prj file.

dsbos commented:
> I think it's just a matter of deciding how to handle exceptions and weird edge cases

I've handled the literal edge case of being on the edge of a time zone, and the method is specified (documented) to return null in the (semi-)exceptional case of not finding a time zone.

Do you think we need to handle anything else?

jamescnowell commented:
I think everything is fine for a rev 1. :shipit:
